proc: cache mapping between goroutine ID and runtime.allgs index

aarzilli opened the pull request:
```
proc: cache mapping between goroutine ID and runtime.allgs index

Caches the mapping between a goroutine ID and its position inside the
runtime.allgs array, this allows speeding up FindGoroutine and makes
commands like 'goroutines -t' not be accidentally quadratic in the
number of goroutines.

```

derekparker commented:
Any reason not to maintain the list of Goroutines as a map regardless? `map[<id>]*G`.

aarzilli commented:
We would have to nuke that every time we resume the target while the map to indexes can survive that.

derekparker commented on pkg/proc/proc.go:
```diff
@@ -588,6 +567,53 @@ func GoroutinesInfo(dbp Process, start, count int) ([]*G, int, error) {
 	return allg, -1, nil
 }
 
+func getAllGs(dbp Process) (uint64, uint64, error) {
```
This function technically only needs a `*BinaryInfo` struct and a `MemoryReadWriter`. Instead of passing in the whole `Process` let's just pass in `dbp.BinInfo()` and `dbp.CurrentThread()`.

derekparker replied:
Also, shouldn't this be the level in which we maintain the allgcache? This gets called from both `GoroutinesInfo` and `FindGoroutine`. We could cache the address of runtime.allg as well. This would offer significant improvement as we won't need to issue a bunch of calls to read process memory, dig through debug info, etc.

derekparker commented on pkg/proc/proc.go:
```diff
@@ -588,6 +567,53 @@ func GoroutinesInfo(dbp Process, start, count int) ([]*G, int, error) {
+func goroutineAtIndex(p Process, i uint64) (*G, error) {
+	allgptr, allglen, err := getAllGs(p)
+	if i >= allglen {
+		return nil, fmt.Errorf("Index out of bounds")
```
nit: s/Index/index/.

derekparker commented on pkg/proc/interface.go:
```diff
@@ -114,6 +114,7 @@ type BreakpointManipulation interface {
 // CommonProcess contains fields used by this package, common to all
 // implementations of the Process interface.
 type CommonProcess struct {
+	goIDToIdx     map[int]uint64
```
Maybe we have a new type here.

Something like (and this is just a sketch):

```
type GoroutineCache struct {
  allgAddr uint64
  allg []*G
  idx  map[int]uint64
}

func (gc *GoroutineCache) AllGAddr() uint64 {
  return gc.allgAddr
}

func (gc *GoroutineCache) AddGoroutine(...) {
  ...
}

func (gc *GoroutineCache) FindGoroutine(...) *G {
 ...
}
```

That way we can abstract it a bit more and callers don't need to know the internals of where the idx map is allocated.

derekparker commented on pkg/proc/proc.go:
```diff
@@ -588,6 +567,53 @@ func GoroutinesInfo(dbp Process, start, count int) ([]*G, int, error) {
+	if i >= allglen {
+		return nil, fmt.Errorf("Index out of bounds")
+	}
+	gvar, err := newGVariable(p.CurrentThread(), uintptr(allgptr+(uint64(i)*uint64(p.BinInfo().Arch.PtrSize()))), true)
```
Can we improve the caching by using a *G we already parsed? I understand that the GoID -> index survives longer because it doesn't depend on our internal cache, but is there a possibility to link them somehow? Like if we already have a parsed the goroutine at this index before and it's just sitting in our allgcache for example, can we just figure out a way for the index cache to return that instead of reading the memory all over again? Since we're directly calling `CurrentThread().ReadMemory()` we're not even getting the benefit of memory caching even if we've already read it before.

aarzilli commented:
I've added a cache for *G objects and removed the GoID -> index cache. I think that if we cache *G objects there is no code path where the other cache gets used.

derekparker commented on pkg/proc/proc.go:
```diff
@@ -588,6 +568,56 @@ func GoroutinesInfo(dbp Process, start, count int) ([]*G, int, error) {
+	return allgptr, allglen, nil
+}
+
+func (gcache *goroutineCache) addGoroutine(i uint64, g *G) {
```
`i` isn't actually used.

derekparker commented on pkg/proc/proc.go:
```diff
@@ -588,6 +568,56 @@ func GoroutinesInfo(dbp Process, start, count int) ([]*G, int, error) {
+	}
+}
+
+func (gcache *goroutineCache) getAllGs(bi *BinaryInfo, mem MemoryReadWriter) (uint64, uint64, error) {
```
nit: The name of this method is a bit misleading. It isn't actually getting Gs it's just returning the pointer to the allg struct and the length.

aarzilli replied:
Changed to getRuntimeAllg, open to alternatives.
